firestore/src/main/java/com/firebase/ui/firestore/paging/FirestorePagingDataAdapter.java

```diff
@@ -0,0 +1,182 @@
+package com.firebase.ui.firestore.paging;
+
+import android.util.Log;
+
+import com.firebase.ui.firestore.SnapshotParser;
+import com.google.firebase.firestore.DocumentSnapshot;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.lifecycle.Lifecycle;
+import androidx.lifecycle.LifecycleObserver;
+import androidx.lifecycle.LiveData;
+import androidx.lifecycle.Observer;
+import androidx.lifecycle.OnLifecycleEvent;
+import androidx.paging.CombinedLoadStates;
+import androidx.paging.LoadState;
+import androidx.paging.PagingData;
+import androidx.paging.PagingDataAdapter;
+import androidx.recyclerview.widget.RecyclerView;
+import kotlin.Unit;
+import kotlin.jvm.functions.Function1;
+
+/**
+ * Paginated RecyclerView Adapter for a Cloud Firestore query.
+ *
+ * Configured with {@link FirestorePagingOptions}.
+ */
+public abstract class FirestorePagingDataAdapter<T, VH extends RecyclerView.ViewHolder>
+        extends PagingDataAdapter<DocumentSnapshot, VH>
+        implements LifecycleObserver {
+
+    private static final String TAG = "FirestorePaging3Adapter";
+    private final Observer<PagingData<DocumentSnapshot>> mDataObserver =
+            new Observer<PagingData<DocumentSnapshot>>() {
+                @Override
+                public void onChanged(@Nullable PagingData<DocumentSnapshot> snapshots) {
+                    if (snapshots == null) {
+                        return;
+                    }
+
+                    submitData(mOptions.getOwner().getLifecycle(), snapshots);
+                }
+            };
+    private FirestorePagingOptions<T> mOptions;
+    private SnapshotParser<T> mParser;
+    private LiveData<PagingData<DocumentSnapshot>> mSnapshots;
+
+    /**
+     * Construct a new FirestorePagingDataAdapter from the given {@link FirestorePagingOptions}.
+     */
+    public FirestorePagingDataAdapter(@NonNull FirestorePagingOptions<T> options) {
+        super(options.getDiffCallback());
+
+        mOptions = options;
+
+        init();
+    }
+
+    /**
+     * Initializes Snapshots and LoadListener
+     */
+    private void init() {
+        mSnapshots = mOptions.getPagingData();
+
+        mParser = mOptions.getParser();
+
+        if (mOptions.getOwner() != null) {
+            mOptions.getOwner().getLifecycle().addObserver(this);
+        }
+
+        addLoadStateListener(new Function1<CombinedLoadStates, Unit>() {
```

Comment:
Ok so the way I see it this function does one of two things:

  - If there's an error, call `onError` 
  - Otherwise we have a big if/else/elseif tree and we eventually map the `(refresh, append)` tuple to a `LoadingState` value and then call `onLoadingStateChanged`
  
  So maybe it would be cleaner if we:
   - First checked for errors and exited early
   - Otherwise used a helper function to map `(refresh, append)` to a single `LoadingState` and then use that return value.

WDYT?